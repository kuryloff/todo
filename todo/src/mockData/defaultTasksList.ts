import { Task } from "../interfaces/todoList.interface";
//@ts-ignore
import { v4 as uuidv4 } from "uuid";

export const defaultTasksList: Task[] = [
  {
    title: "Tasks creation management",
    id: uuidv4(),
    description: "Add and remove tasks",
    status: 0,
    subtasks: [
      {
        id: uuidv4(),
        title: "Add tasks",
        status: 0,
      },
      {
        id: uuidv4(),
        title: "Remove tasks",
        status: 0,
      },
    ],
  },
  {
    title: "Additional display info",
    id: uuidv4(),
    description: "Display a count of remaining incomplete tasks",
    status: 0,
    subtasks: [
      {
        id: uuidv4(),
        title: "Show counter",
        status: 1,
      },
      {
        id: uuidv4(),
        title: "Apply counter function",
        status: 0,
      },
    ],
  },
  {
    title: "Local data saving",
    id: uuidv4(),
    description: "Data should persisted using some form of browser storage",
    status: 0,
    subtasks: [
      {
        id: uuidv4(),
        title: "Create storage",
        status: 0,
      },
      {
        id: uuidv4(),
        title: "Update storage functionality",
        status: 0,
      },
    ],
  },
  {
    title: "Tasks display",
    id: uuidv4(),
    description: "Should display as a list of Task cards",
    status: 0,
    subtasks: [
      { id: uuidv4(), title: "Create Task Card", status: 0 },
      { id: uuidv4(), title: "Arrange display", status: 0 },
    ],
  },
  {
    title: "testing",
    id: uuidv4(),
    description: "Should have unit tests",
    status: 0,
    subtasks: [],
  },
  {
    title: "Task completion",
    id: uuidv4(),
    description:
      "Should be able to mark a task status as complete by clicking a checkbox",
    status: 0,
    subtasks: [{ id: uuidv4(), title: "Create update functions", status: 0 }],
  },
  {
    title: "Update",
    id: uuidv4(),
    description: "Should be able to update all criteria of a task",
    status: 0,
    subtasks: [
      {
        id: uuidv4(),
        title: "Should show title as crossed out when complete",
        status: 0,
      },
      { id: uuidv4(), title: "Arrange display", status: 0 },
    ],
  },
];
