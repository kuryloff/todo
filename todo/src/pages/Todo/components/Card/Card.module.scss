@import "src/styles/mixins";
.container,
.containerDone{
  //width: 100%;
  font-size: 16px;
  border: 1px solid #a6b1ef;
  border-radius: 5px;
  margin: 5px;
  padding: 5px;
  box-shadow: 5px 4px 15px -4px #8a8989;
  background-color: #f2f8f8;
  position: relative;
}

.containerDone, .subtaskDone{
  text-decoration: line-through;
}

.title{
  margin-left: 5px;
  font-weight: 600;
}

.header{
  @include flexStartCenter;
}

.description{
  margin-left: 20px;
  font-size: 12px;
 }

.subtasksContainer{
  margin-left: 30px;
}

.subtask, .subtaskDone{
  font-size: 13px;
  @include flexStartCenter;
}

.controls{
  position: absolute;
  right: 5px;
  top: 0px;
  border: none;
  border-radius: 3px;

  button{
    margin: 5px;
    color: white;
    border: none;
    border-radius: 3px;
    font-weight: 600;
  }
  :hover{
    box-shadow: 5px 4px 15px -4px #8a8989;
  }
  button:nth-child(1){
    background-color: orange;

  }
  button:nth-child(2){
    background-color: red;
  }
}

.blur{
  filter: blur(2px);
}