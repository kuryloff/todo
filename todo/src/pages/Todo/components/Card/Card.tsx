import React from "react";
import { Status, Task } from "../../../../interfaces/todoList.interface";
import styles from "./Card.module.scss";

interface Props {
  task: Task;
  onSubtaskChange: (e: any, a: number, i: number) => void;
  onTaskChange: (e: any, i: number) => void;
  deleteTask: (i: number) => void;
  handleEditItem: (i: number) => void;
  taskIndex: number;
  editIndex: number;
  editMode: boolean;
  addMode: boolean;
}
const Card = ({
  task,
  onSubtaskChange,
  taskIndex,
  editIndex,
  onTaskChange,
  deleteTask,
  handleEditItem,
  editMode,
  addMode,
}: Props) => {
  return (
    <div
      className={editIndex !== taskIndex && editMode ? styles.blur : ""}
      data-testid={task.id}
    >
      <div
        className={
          task.status === Status.inList
            ? styles.container
            : styles.containerDone
        }
      >
        <div className={styles.controls}>
          <button onClick={() => handleEditItem(taskIndex)}>Edit</button>
          <button onClick={() => deleteTask(taskIndex)}>Delete</button>
        </div>
        <div className={styles.header}>
          <input
            type="checkbox"
            checked={Boolean(task.status)}
            onChange={(e) => onTaskChange(e, taskIndex)}
          />
          <span className={styles.title}>{task.title}</span>
        </div>
        <div className={styles.description}>
          <span>{task.description}</span>
        </div>
        {task.subtasks && task.subtasks?.length > 0 && (
          <div className={styles.subtasksContainer}>
            {task.subtasks.map((subtask, index) => (
              <div
                className={
                  subtask.status === Status.inList
                    ? styles.subtask
                    : styles.subtaskDone
                }
                key={subtask.id}
              >
                <input
                  type="checkbox"
                  checked={Boolean(subtask.status)}
                  onChange={(e) => onSubtaskChange(e, taskIndex, index)}
                />
                <span key={subtask.id} className={styles.title}>
                  <small>{subtask.title}</small>
                </span>
              </div>
            ))}
          </div>
        )}
      </div>
    </div>
  );
};

export default Card;
