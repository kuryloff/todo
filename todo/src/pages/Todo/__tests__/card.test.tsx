import { render, screen, cleanup } from "@testing-library/react";
import React from "react";
import Card from "../components/Card/Card";
import { defaultTasksList } from "../../../mockData/defaultTasksList";
import renderer from "react-test-renderer";

afterEach(() => {
  cleanup();
});
// @ts-ignore
test("should render Card Component with subtasks", () => {
  render(
    <Card
      editMode={true}
      addMode={false}
      editIndex={1}
      task={defaultTasksList[0]}
      taskIndex={0}
      key={"string"}
      onTaskChange={() => console.log("change")}
      onSubtaskChange={() => console.log("change")}
      deleteTask={() => console.log("change")}
      handleEditItem={() => console.log("change")}
    />
  );
  const card = screen.getByTestId(defaultTasksList[0].id);
  expect(card).toBeInTheDocument();
  expect(card).toHaveTextContent(defaultTasksList[0].title);
  expect(card).toHaveTextContent(defaultTasksList[0].description);
  // @ts-ignore
  expect(card).toHaveTextContent(defaultTasksList[0]?.subtasks[0]?.title);
  // @ts-ignore
  expect(card).toHaveTextContent(defaultTasksList[0]?.subtasks[1]?.title);
  expect(card).toContainHTML("small");
});

test("should render Card Component without subtasks", () => {
  render(
    <Card
      editMode={true}
      addMode={false}
      editIndex={1}
      task={defaultTasksList[4]}
      taskIndex={0}
      key={"string"}
      onTaskChange={() => console.log("change")}
      onSubtaskChange={() => console.log("change")}
      deleteTask={() => console.log("change")}
      handleEditItem={() => console.log("change")}
    />
  );
  const card = screen.getByTestId(defaultTasksList[4].id);
  expect(card).toBeInTheDocument();
  expect(card).toHaveTextContent(defaultTasksList[4].title);
  expect(card).toHaveTextContent(defaultTasksList[4].description);
  expect(card).not.toContainHTML("small");
});
//todo need to update snapshot for running tests due to uuid
test("Card snapshot match", () => {
  const dom = renderer
    .create(
      <Card
        editMode={true}
        addMode={false}
        editIndex={1}
        task={defaultTasksList[0]}
        taskIndex={0}
        key={defaultTasksList[0].id}
        onTaskChange={() => console.log("change")}
        onSubtaskChange={() => console.log("change")}
        deleteTask={() => console.log("change")}
        handleEditItem={() => console.log("change")}
      />
    )
    .toJSON();
  expect(dom).toMatchSnapshot();
});
