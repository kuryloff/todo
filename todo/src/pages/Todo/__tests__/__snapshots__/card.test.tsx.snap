// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card snapshot match 1`] = `
<div
  className="blur"
  data-testid="fa032c90-0222-48e4-983b-06ab03eeecd3"
>
  <div
    className="container"
  >
    <div
      className="controls"
    >
      <button
        onClick={[Function]}
      >
        Edit
      </button>
      <button
        onClick={[Function]}
      >
        Delete
      </button>
    </div>
    <div
      className="header"
    >
      <input
        checked={false}
        onChange={[Function]}
        type="checkbox"
      />
      <span
        className="title"
      >
        Tasks creation management
      </span>
    </div>
    <div
      className="description"
    >
      <span>
        Add and remove tasks
      </span>
    </div>
    <div
      className="subtasksContainer"
    >
      <div
        className="subtask"
      >
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="title"
        >
          <small>
            Add tasks
          </small>
        </span>
      </div>
      <div
        className="subtask"
      >
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="title"
        >
          <small>
            Remove tasks
          </small>
        </span>
      </div>
    </div>
  </div>
</div>
`;
